### PR TITLE
Knobs addon: new knob type `button`

### PR DESCRIPTION
**Pull request by derrickpelletier**
Issue: #1914

## What I did
- added button knob to the knobs addon so users don't need to include buttons in their stories
- added ability to hide the label, for button type (and maybe future types?) to render nicer.
- added `button` example to `cra-kitchen-sink` and `vue-kitchen-sink` with-knobs examples
- added `button` usage to `addons/knobs/README.md`

## How to test
- acccess either vue or cra kitchen sink examples

### Is this testable with jest or storyshots?
- I don't think so, unless I am missing something.

### Does this need a new example in the kitchen sink apps?
- done

### Does this need an update to the documentation?
- done

### CODE REVIEW

File: addons/knobs/src/react/WrapStory.js [hunk 1]
```diff
@@ -45,6 +45,10 @@ export default class WrapStory extends React.Component {
     const { knobStore, storyFn, context } = this.props;
     // Update the related knob and it's value.
     const knobOptions = knobStore.get(name);
+
+    // if the knob is a button, a change event should call the callback
+    if (knobOptions.callback) knobOptions.callback();
```
Comment (Hypnosphi) [inline addons/knobs/src/react/WrapStory.js]: Let's use a separate method for that. There're no "changes" here actually

File: addons/knobs/README.md [hunk 2]
```diff
@@ -231,6 +231,18 @@ const value = date(label, defaultValue);
...
+### button
+
+Allows you to include a button and associated handler.
+
+```js
+import { button } from '@storybook/addon-knobs';
+
+const label = 'Do Something';
+const handler = () => doSomething('foobar');
+const value = button(label, handler);
```
Comment (Hypnosphi) [inline addons/knobs/README.md]: What can you do with that `value`?

**Review by Hypnosphi — CHANGES_REQUESTED** [review]

File: addons/knobs/src/react/WrapStory.js (hunk 1 above)
Comment (derrickpelletier) [reply addons/knobs/src/react/WrapStory.js]: Fair point. I added click handlers that live alongside the change handlers.
Both `onClick` and `onChange` are passed to each `InputType` now, but it's up to that particular `InputType` (in this case, `ButtonType`) to decide how to use each.

File: addons/knobs/README.md (hunk 2 above)
Comment (derrickpelletier) [reply addons/knobs/README.md]: Addressed, removed the assignment as it was unnecessary.

**Review by Hypnosphi — APPROVED** [review]

**Comment by Hypnosphi** [comment]
Code LGTM. Still not sure this thing belongs to addon-knobs semantically

**Comment by derrickpelletier** [comment]
Ok, who decides how to move forward then? @danielduan?

I can see where you're coming from, because it's functionally a bit different than inputs—but semantically it makes sense to me as a way to control or trigger behaviour in a story.

**Comment by danielduan** [comment]
@ndelangen @shilman thoughts?

I definitely see why there's a need for this, but maybe this isn't a "knob"? perhaps it should be its own addon?

File: addons/knobs/README.md [hunk 3]
```diff
@@ -231,6 +231,18 @@ const value = date(label, defaultValue);
...
+### button
+
+Allows you to include a button and associated handler.
+
+```js
+import { button } from '@storybook/addon-knobs';
+
+const label = 'Do Something';
+const handler = () => doSomething('foobar');
+button(label, handler);
```
Comment (ndelangen) [inline addons/knobs/README.md]: The way this is used, is quite different from how the other knobs are used. This confuses me, and I'd assume other as well. Having a fully 'working' real-world example would help a lot understanding this.

File: examples/cra-kitchen-sink/src/stories/index.js [hunk 4]
```diff
@@ -110,6 +111,8 @@ storiesOf('Button', module)
     const salutation = nice ? 'Nice to meet you!' : 'Leave me alone!';
     const dateOptions = { year: 'numeric', month: 'long', day: 'numeric' };
 
+    button('Arbitrary action', action('You clicked it!'));
```
Comment (ndelangen) [inline examples/cra-kitchen-sink/src/stories/index.js]: This example is rather meaningless right? calling `action` from a button inside the knobs panel, won't be useful, or am I missing something here?

Could you add an example that would help understand the true use-case of this change? 👍

**Review by ndelangen — CHANGES_REQUESTED** [review]
Code is good!

I'm struggling understanding the use-case, and I need better examples in order to let this merge.

As it is right now, I wouldn't know how to use it for a real purpose.

File: examples/cra-kitchen-sink/src/stories/index.js (hunk 4 above)
Comment (derrickpelletier) [reply examples/cra-kitchen-sink/src/stories/index.js]: Totally reasonable

Comment (derrickpelletier) [reply examples/cra-kitchen-sink/src/stories/index.js]: So I made an example in the CRA for a case where i've normally included a button _inside_ the story itself to trigger something like loading some external data through a render callback component, or generating a set of random data. These are the kind of use cases I'd have in mind for something like this. 

I haven't thought too hard about it, there could be better ways to do it. 

If you think that's not a good example, or not within knobs paradigm—totally fine. Don't have my heart set on it :)

File: examples/cra-kitchen-sink/src/stories/index.js [hunk 5]
```diff
@@ -60,6 +61,23 @@ const InfoButton = () => (
...
+    super();
+    this.state = { items: [] };
+  }
+
+  loadItems() {
+    setTimeout(() => this.setState({ items: ['pencil', 'pen', 'eraser'] }), 1500);
+  }
+
+  render() {
+    button('Load the items', () => this.loadItems());
```
Comment (ndelangen) [inline examples/cra-kitchen-sink/src/stories/index.js]: Shouldn't this be in the constructor?

**Review by ndelangen — CHANGES_REQUESTED** [review]

File: examples/cra-kitchen-sink/src/stories/index.js (hunk 5 above)
Comment (ndelangen) [reply examples/cra-kitchen-sink/src/stories/index.js]: Very clear example by the way, Thanks!

Comment (derrickpelletier) [reply examples/cra-kitchen-sink/src/stories/index.js]: HMR kills it when it is in the constructor. The button stays present in the knob panel but the instance of the example component is new, so the actual component tied to the knob was unmounted.

Comment (ndelangen) [reply examples/cra-kitchen-sink/src/stories/index.js]: Ok, if this is the best solution, that's fine.

Can you take a look at why the CI is disapproving of this PR?

Comment (ndelangen) [reply examples/cra-kitchen-sink/src/stories/index.js]: I restarted the build tasks

**Review by ndelangen — APPROVED** [review]

**Comment by ndelangen** [comment]
Great job @derrickpelletier ! Thank you!

File: addons/knobs/README.md (hunk 3 above)
Comment (dperetti) [reply addons/knobs/README.md]: Looking at the documentation as of today, I don't get nothin' 😆. How is this supposed to be used ? 🤔

**Pull request merged**